binderhub/tests/test_utils.py

```diff
@@ -0,0 +1,25 @@
+from binderhub import utils
+
+
+def test_rendezvous_rank():
+    # check that a key doesn't move if its assigned bucket remains but the
+    # other buckets are removed
+    key = "k1"
+    first_round = utils.rendezvous_rank(["b1", "b2", "b3"], key)
+    second_round = utils.rendezvous_rank([first_round[0][1], first_round[1][1]], key)
```

Comment:
I don't think that keys have to remain associated to the same bucket if a new one is added.

Imagine you have three buckets that split the space like this: `--->b1------->b2------>b3--`. Let's assume that the key `k` is placed like this `--->b1--k---->b2------>b3--`. This means it will get placed in `b2`. Now we add `b4` and it just happens to get placed between `k` and `b2` like so `--->b1--k-->b4->b2------>b3--`. Now `k` will be placed on `b4`, not `b2` any more.

I think it makes sense that this is true because if it wasn't then no keys would ever get assigned to a new bucket when one is added. Starting with a situation with only one bucket `b1`, then any possible key `kX` would get assigned to `b1`. Now we add `b2`. Now we want about half of all the keys `kX` to be assigned to `b2` and half to `b1`. So some keys have to move.

Reply:
We agree, I don't think that keys will remain either if a new bucket is added.

I think there is symmetry still that allows us to conclude what the comment suggested - that we actually successfully test both removing and adding. If we would have started calculating 90 buckets and then calculated 90+10 and slimmed the second ranking, we would have tested for adding, but now we calculate 90+10 first and then for 90 etc.

Reply:
```python
    # verify ranking order is retained after buckets
    # are removed, but also if they are added, which
    # follows from symmetry

    key = "repo1"
    buckets = ["bucket_"+str(i) for i in range(90)]
    extra_buckets = ["extra_bucket_"+str(i) for i in range(10)]

    ranking_1 = utils.rendezvous_rank(buckets + extra_buckets, key)
    for extra_bucket in extra_buckets:
        ranking_1.remove(extra_bucket)

    ranking_2 = utils.rendezvous_rank(buckets, key)

    assert ranking_1 == ranking_2
```

Reply:
Resolving this conversation as I added a new test implementing your suggestion. If we need to lets start a new conversation connected to a line of that test.